txnkv/transaction/pessimistic.go

```diff
@@ -147,152 +165,384 @@ func (action actionPessimisticLock) handleSingleBatch(c *twoPhaseCommitter, bo *
 		sender := locate.NewRegionRequestSender(c.store.GetRegionCache(), c.store.GetTiKVClient())
 		startTime := time.Now()
 		resp, err := sender.SendReq(bo, req, batch.region, client.ReadTimeoutShort)
-		reqDuration := time.Since(startTime)
+		diagCtx.reqDuration = time.Since(startTime)
+		diagCtx.sender = sender
 		if action.LockCtx.Stats != nil {
-			atomic.AddInt64(&action.LockCtx.Stats.LockRPCTime, int64(reqDuration))
+			atomic.AddInt64(&action.LockCtx.Stats.LockRPCTime, int64(diagCtx.reqDuration))
 			atomic.AddInt64(&action.LockCtx.Stats.LockRPCCount, 1)
 		}
 		if err != nil {
 			return err
 		}
-		regionErr, err := resp.GetRegionError()
-		if err != nil {
-			return err
-		}
-		if regionErr != nil {
-			// For other region error and the fake region error, backoff because
-			// there's something wrong.
-			// For the real EpochNotMatch error, don't backoff.
-			if regionErr.GetEpochNotMatch() == nil || locate.IsFakeRegionError(regionErr) {
-				err = bo.Backoff(retry.BoRegionMiss, errors.New(regionErr.String()))
-				if err != nil {
-					return err
-				}
+
+		if action.lockWaitMode == kvrpcpb.PessimisticWaitLockMode_RetryFirst {
+			finished, err := action.handlePessimisticLockResponseRetryFirstMode(c, bo, &batch, mutations, resp, &diagCtx)
+			if err != nil {
+				return err
+			}
+			if finished {
+				return nil
 			}
-			same, err := batch.relocate(bo, c.store.GetRegionCache())
+		} else if action.lockWaitMode == kvrpcpb.PessimisticWaitLockMode_LockFirst {
+			finished, retryMutations, err := action.handlePessimisticLockResponseLockFirstMode(c, bo, &batch, mutations, resp, &diagCtx)
```

Comment:
Maybe we could record how many pessimistic RPC requests are sent before finishing, and also how many resolving operations are triggered.
If the whole processing takes several seconds and tikv keeps telling the kv-client it needs not to do the resolving, does it mean that this request is always at the back of the waiting queue which could be unexpected?

Reply:
Yes. But since TiKV uses a priority queue based on the transaction's start_ts, I think it's impossible.